Fix for DataFrame.hist() with by- and weights-keyword

Twizzledrizzle opened the pull request:
will make the following work

``` python
import numpy as np
import pandas as pd
d = {'one' : ['A', 'A', 'B', 'B', 'C'],
     'two' : [4., 3., 2., 1., np.nan],
     'three' : [10., 8., np.nan, 5., 7.]}     
df = pd.DataFrame(d)
df.hist('two', by='one', weights='three', bins=range(0, 10))
# or
df.hist('two', by=df.one.values, weights='three', bins=range(0, 10))
```

nosterlu committed:
Fix for DataFrame.hist() with by- and weights-keyword

will make the following work

import pandas as pd
d = {'one' : ['A', 'A', 'B', 'C'],
     'two' : [4., 3., 2., 1.],
     'three' : [10., 8., 5., 7.]}
df = pd.DataFrame(d)
df.hist('two', by='one', weights='three', bins=range(0, 10))

Twizzledrizzle commented:
Since my poor git skills, this is a continuation for
https://github.com/pydata/pandas/pull/11028

upgraded to work with the newest pandas release

I am at loss writing a test for this. @jreback I could not find the place you recommended. The place where weights are tested right now seems to be test_generic.py?

However I am unsure of how to construct a test the best way, I have never written one, so pointers are very welcome!

nosterlu committed:
Fix for handling both ndarray and series.

Also fix if NaN's make an entire group empty

nosterlu committed:
Fix for if weights is supplied by an array instead of the column name in the DataFrame

Twizzledrizzle commented:
Note, I added the column 'weight' to the supplied DataFrame if an array was supplied instead of the column name... Can this be done in a nicer way? I need it in the groupby logic.

if a column is named 'weight' it will be overwritten... :(

TomAugspurger commented:
@Twizzledrizzle no worries about the git snafu, we've all been there. People are usually in the [chat](https://gitter.im/pydata/pandas) if you need help walking through git.

I think Jeff was talking about the logic in `DataFrame.sample` in `pandas/core/frame.py`. I haven't had a chance to look at your code to see if it's similar enough to refactor into a common function.

For the tests you can look to see how similar tests are done in `pandas/tests/test_graphics.py`. Basically you call the function your testing and assert that something about the result equals an expected value that you hardcode in the test.

I'll take a look later, but it won't be until next week sometime.

nosterlu committed:
Fixed bug where weights were not used, since not included in **kwds anymore

Also made the logic better when using weights without group by by aligning NaN's, and finding out if weights is supplied by column or by array.

Twizzledrizzle commented:
Thanks @TomAugspurger I will take a look.

A lot of commits right now, finding things while working through the travis errors... I hope the server does not crash!

TomAugspurger commented:
Yeah, we'll come up with a better way to solve that weights column.

FYI you can run the tests locally, will be quicker than waiting on Travis. Make sure you have nose installed (`pip install nose` or `conda install nose) and then run the tests with `nosetests test_graphics` from the `pandas/tests` directory.

Twizzledrizzle commented:
I have python 2.7 and every tests seems to work fine. Why could python 3.4 fail on 

```
nosetests test_data:TestGoogle.test_get_multi1
```

Does not seem to have anything to do with my pull request?

jreback commented on pandas/tools/plotting.py:
```diff
@@ -2770,9 +2770,10 @@ def plot_group(group, ax):
-               xrot=None, ylabelsize=None, yrot=None, ax=None, sharex=False,
-               sharey=False, figsize=None, layout=None, bins=10, **kwds):
+def hist_frame(data, column=None, weights=None, by=None, grid=True,
+               xlabelsize=None, xrot=None, ylabelsize=None, yrot=None, ax=None,
```
don't change the order of parameters, just add at th eend

jreback commented on pandas/tools/plotting.py:
```diff
@@ -2781,6 +2782,8 @@ def hist_frame(data, column=None, by=None, grid=True, xlabelsize=None,
     column : string or sequence
         If passed, will be used to limit data to a subset of columns
+    weights : string or sequence
+        If passed, will be used to weight the data
```
same here

jreback commented on pandas/tools/plotting.py:
```diff
@@ -2810,19 +2813,35 @@ def hist_frame(data, column=None, by=None, grid=True, xlabelsize=None,
+        if isinstance(weights, str):
+            weights = data[weights]
+        if isinstance(weights, np.ndarray) == False:
+            weights = weights.values
```
this code is pretty close to ``DataFrame.sample`` need to consolidate into a function somewhere

jreback commented on pandas/tools/plotting.py:
```diff
@@ -2942,12 +2962,26 @@ def grouped_hist(data, column=None, by=None, ax=None, bins=50, figsize=None,
+            # for both data sets
+            if isinstance(weights, np.ndarray) == False:
+                weights = weights.values
+            inx_na |= (np.isnan(weights))
```
same here this is pretty duplicative

jreback commented:
need tests

jreback commented:
@Twizzledrizzle ignore that goggle multi failure....its an unreliable test on travis

nosterlu committed:
changed weights= to end of functions

Uses dropna(subset=...) to delete where nan's over the columns supplied
Also doing this in the beginning so we do not have to duplicate this logic

Twizzledrizzle commented:
Made the code more readable now I think, I hope it looks better.

Regarding tests, are we talking tests like this? In test_graphics_others.py? (Looks like some other tests are done one the hist functionality there).

ax = _check_plot_works(df.hist, column='two', by='one', weights='three', bins=range(0, 10))

nosterlu committed:
Added the first test (in my life)

Twizzledrizzle commented:
Oh also, as a side, the new function df.plot.hist() will miss a lot of this functionality.

df.plot.hist() seem to have problems using the by- keyword. Also missing column keyword. I tried working with the code, but it was more complex than the one I have been editing. I could not get my head around making the by-keyword work. 

Should I open a discussion around this perhaps in a separate thread?

Twizzledrizzle commented on pandas/tools/plotting.py:
```diff
@@ -2807,17 +2808,37 @@ def hist_frame(data, column=None, by=None, grid=True, xlabelsize=None,
+    subset_cols_drop_nan = []
+    if weights is not None:
+        if isinstance(weights, np.ndarray):
+            # weights supplied as an array instead of a part of the dataframe
```
Hmm, this will not work if weights is a >1 dimensional ndarray... Need to think on this

jreback commented:
@TomAugspurger status of this?

Twizzledrizzle commented:
I am awaiting with interest what will happen with

https://github.com/pydata/pandas/pull/8018

before putting to much time in this pull request.

jreback commented:
@TomAugspurger status of this

jreback commented:
@TomAugspurger status?

TomAugspurger commented:
I think waiting on #8018, which will conflict with the case where `by` is not None here.

jreback commented:
status?

jreback commented:
@TomAugspurger can this be independent #8018 ?

jreback commented:
@TomAugspurger ?

jreback commented:
closing as stale
